src/EditorFeatures/Core/Tagging/AbstractAsynchronousTaggerProvider.TagSource_ProduceTags.cs

```diff
@@ -461,74 +408,21 @@ private Task ProduceTagsAsync(TaggerContext<TTag> context)
-                    var asyncToken = this._asyncListener.BeginAsyncOperation(nameof(ProcessNewTagTrees));
-                    _ = Task.Run(async () =>
-                    {
-                        await this.ThreadingContext.JoinableTaskFactory.SwitchToMainThreadAsync(cancellationToken);
-                        UpdateStateAndReportChanges(newTagTrees, bufferToChanges, newState, initialTags);
-                    }, CancellationToken.None).CompletesAsyncOperation(asyncToken); // TODO: What should the cancellation behavior be here? passing CancellationToken.None for now
-                }
-                else
-                {
-                    // Otherwise report back on the foreground to update the state and let our clients know about the
-                    // change.  This will go to the end of the foreground processing queue.  This will normally process
-                    // quickly once VS is loaded, but it may take some time initially when VS is loading and the UI
-                    // thread is highly occupied.  This helps ensure that we don't oversaturate the UI during a very
-                    // contended period of time.
-                    RegisterNotification(() => UpdateStateAndReportChanges(
-                        newTagTrees, bufferToChanges, newState, initialTags),
-                        delay: 0,
-                        cancellationToken: cancellationToken);
-                }
-            }
+                await this.ThreadingContext.JoinableTaskFactory.SwitchToMainThreadAsync(cancellationToken);
 
-            private void UpdateStateAndReportChanges(
-                ImmutableDictionary<ITextBuffer, TagSpanIntervalTree<TTag>> newTagTrees,
-                Dictionary<ITextBuffer, DiffResult> bufferToChanges,
-                object newState,
-                bool initialTags)
-            {
                 this.AssertIsForeground();
 
                 // Now that we're back on the UI thread, we can safely update our state with
-                // what we've computed.  There is no concern with race conditions now.  For 
-                // example, say that another change happened between the time when we 
-                // registered for UpdateStateAndReportChanges and now.  If we processed that
-                // notification (on the UI thread) first, then our cancellation token would 
-                // have been triggered, and the foreground notification service would not 
-                // call into this method. 
-                // 
-                // If, instead, we did get called into, then we will update our instance state.
-                // Then when the foreground notification service runs RecomputeTagsForeground
-                // it will see that state and use it as the new basis on which to compute diffs
-                // and whatnot.
+                // what we've computed.
                 this.CachedTagTrees = newTagTrees;
-                this.AccumulatedTextChanges = null;
```

Comment:
this moved up.  now that computation is not cancellable, we don't have to wait till the computation ends to null this out.  we can do it when we issue the computation request.